use property map with BOOST_NO_RTTI and gcc

dynamic_property_map.hpp fails to compile with BOOST_NO_RTTI defined using clang or gcc.

Example program:
`
#define BOOST_NO_RTTI
#include <boost/property_map/dynamic_property_map.hpp>

int main()
{
}
`
Error message with gcc 10:
boost/property_map/dynamic_property_map.hpp: In member function 'void boost::detail::dynamic_property_map_adaptor<PropertyMap>::do_put(const boost::any&, const boost::any&, mpl_::bool_<true>)':
/opt/wandbox/boost-1.70.0/gcc-head/include/boost/property_map/dynamic_property_map.hpp:150:25: error: no match for 'operator==' (operand types are 'const type_info' {aka 'const boost::typeindex::detail::ctti_data'} and 'const std::type_info')
  150 |     if (in_value.type() == typeid(value_type)) {
      |         ~~~~~~~~~~~~~~~ ^~ ~~~~~~~~~~~~~~~~~~
      |                      |     |
      |                      |     const std::type_info
      |                      const type_info {aka const boost::typeindex::detail::ctti_data}

The problem seems to be that typeid id used instead of boost::typeindex::type_id.

--- a/include/boost/property_map/dynamic_property_map.hpp
+++ b/include/boost/property_map/dynamic_property_map.hpp
@@ -26,6 +26,7 @@
 #include <boost/type_traits/is_convertible.hpp>
 #include <boost/mpl/bool.hpp>
 #include <boost/type.hpp>
+#include <boost/type_index.hpp>
 #include <boost/smart_ptr.hpp>
 #include <exception>
 #include <map>
@@ -149,7 +150,7 @@ class dynamic_property_map_adaptor : public dynamic_property_map
     using boost::put;
 
     key_type key_ = any_cast<key_type>(in_key);
-    if (in_value.type() == typeid(value_type)) {
+    if (in_value.type() == boost::typeindex::type_id<value_type>()) {
       put(property_map_, key_, any_cast<value_type>(in_value));
     } else {
       //  if in_value is an empty string, put a default constructed value_type.

--- a/test/Jamfile.v2
+++ b/test/Jamfile.v2
@@ -21,6 +21,7 @@ test-suite property_map
   : [ compile property_map_cc.cpp ]
     [ run     compose_property_map_test.cpp ]
     [ run     dynamic_properties_test.cpp ]
+    [ run     dynamic_properties_test.cpp : : : <define>BOOST_NO_RTTI=1 : dynamic_properties_no_rtti_test ]
     [ run     function_property_map_test.cpp ]
     [ run     transform_value_property_map_test.cpp ]
   ;

--- a/test/suppressions.txt
+++ b/test/suppressions.txt
@@ -1,1 +1,2 @@
 leak:dynamic_properties_test
+leak:dynamic_properties_no_rtti_test

